ide-plugin/src/org/jetbrains/plugins/ruby/ancestorsextractor/support/Support.kt

```diff
@@ -0,0 +1,113 @@
+
+        PrintWriter(proc.outputStream, true).use { it.println(toExec) }
+        // free buffers
+        val ret = RailsConsoleExecutionResult(
+                proc.inputStream.bufferedReader().readText(), proc.errorStream.bufferedReader().readText())
+
+        proc.waitFor()
+        return ret
+    } catch(e: Exception) {
+        throw Exception("Error while launching rails console")
+    }
+}
+
+/**
+ * Extract information left by [rubyCode] in [tempJSONFilPath]
+ * @param clazz which kind of information [rubyCode] left in [tempJSONFilPath]
+ * @param tempJSONFilPath path to temp file where [rubyCode] left information which
+ *        can be converted from JSON to [clazz]
+ * @param rubyCode Your ruby code which should leave some JSON information in [tempJSONFilPath]
+ */
+@Throws(Exception::class)
+fun <T> extractFromRubyOnRailsConsole(project: Project, clazz: Class<T>,
+                                      tempJSONFilPath: String, rubyCode: String): T {
+    val projectDirPath = project.basePath ?: throw Exception("Seems that project is default. " +
+            "Quote from com.intellij.openapi.project.Project#getBasePath JavaDoc")
+
+    val railsConsoleExecutionResult = runRailsConsole(projectDirPath, rubyCode)
+
+    val file = File(tempJSONFilPath)
+    val ret = try {
+        file.inputStream().bufferedReader().use {
+            Gson().fromJson(it.readLine(), clazz)
+        }
+    } catch (ex: IOException) {
+        null
+    }
+    return ret ?: throw Exception("""
+            |Error occurred either in following Ruby code:
+            |$rubyCode
+
+            |stdout of this Ruby code execution:
+            |${railsConsoleExecutionResult.stdout}
+
+            |stderr of this Ruby code execution:
+            |${railsConsoleExecutionResult.stderr}
+
+            |either while trying to read data from JSON left by Ruby
+        """.trimMargin())
+}
+
+/**
+ * Provides temp file paths
+ * @param initTempFileName lambda where you can generate temp file name
+ */
+class TempFilePathProvider(private val initTempFileName: () -> String) {
+    private var _path: String? = null
+    /**
+     * Path to temp file.
+     * @see removeTempFileIfExistsAndForgetAboutIt
+     */
+    val path: String
+        get() {
+            val pathLocal = _path ?: initTempFileName()
+            _path = pathLocal
+            return pathLocal
+        }
+
+    /**
+     * Remove temp file if exists. And forget about that temp file path, that means that next [path] call
+     * will call [initTempFileName] once more to get new temp file path and since that moment will return
+     * it until next [removeTempFileIfExistsAndForgetAboutIt] call.
+     */
+    fun removeTempFileIfExistsAndForgetAboutIt() {
+        val pathLocal = _path
+        if (pathLocal != null) {
+            File(pathLocal).delete()
+            _path = null
+        }
+    }
+}
+
+/**
+ * Singleton which keeps all temp files paths
+ */
+object TempFiles {
```

Comment:
it will create a json file on every run. Please, consider https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/lazy.html

Reply:
TempFilePathProvider works like Kotlin's lazy with ability of value regeneration